### ISSUE TITLE
Vertex tool calls with roundtrips not working

### ISSUE THREAD

**Issue by sumitstranger** [vercel/ai#2356]
### Description

When the tool is invoked and the value of maxToolRoundtrips is greater than 0, the following error occurs:

```
 ⨯ ClientError: [VertexAI.ClientError]: got status: 400 Bad Request. {"error":{"code":400,"message":"Invalid JSON payload received. Unknown name \"response\" at 'contents[2].parts[0].function_response': Proto field is not repeating, cannot start list.","status":"INVALID_ARGUMENT","details":[{"@type":"type.googleapis.com/google.rpc.BadRequest","fieldViolations":[{"field":"contents[2].parts[0].function_response","description":"Invalid JSON payload received. Unknown name \"response\" at 'contents[2].parts[0].function_response': Proto field is not repeating, cannot start list."}]}]}}
    at async continueConversation (./src/app/chat/actions.tsx:28:35)
```

SDK Versions:
```
    "@ai-sdk/google-vertex": "^0.0.16",
     "ai": "^3.2.32",
      "next": "^14.2.1",
 ```

### Code example

File 'src/app/actions.tsx'

'use server';

import { generateText } from 'ai';
import { createVertex } from '@ai-sdk/google-vertex';
import { z } from 'zod';

export interface Message {
    role: 'user' | 'assistant';
    content: string;
}

const getOfferList = async (query: string) => {
//....
return [
//{....},
//{....}
];
};

export async function continueConversation(history: Message[]) {
    'use server';

    const { text, toolResults } = await generateText({
        maxToolRoundtrips: 3,
        model: createVertex({
            project: '_PROJECT_ID',
            location: '_LOCATION_'
        })('gemini-1.5-pro'),
        system: 'You are a helpful assistant',
        messages: history,

        tools: {
            getDiscountOffer: {
                description: 'Get Offers on user request',
                parameters: z.object({
                    query: z.string().describe('User Request for Offers'),
                }),
                execute: async ({ query }) => {
                    const offerList = (await getOfferList(query)).map((offer: any) => offer.content);
                    return offerList;
                },
            },
        },
    });

    return {
        messages: [
            ...history,
            {
                role: 'assistant' as const,
                content:
                    text || toolResults.map(toolResult => toolResult.result).join('\n'),
            },
        ],
    };
}

**Comment by neoromantic** [comment]
I struggle with same problem. 

I have code that works OK with OpenAI, but if i switch to Gemini, i get same problem with tool calling and roundtrips. 

Here's how its with me:

```js
export async function POST(req: Request) {
  const { messages } = await req.json()

  const assembledMessages = [
    {
      role: "system",
      content: promptInstance,
    },
    ...messages,
  ]

  const result = await streamText({
    model: vertex("gemini-1.5-pro"),
    // model: openai("gpt-4o"),
    messages: convertToCoreMessages(assembledMessages),
    tools: {
      getReports: tool({
        description: `....`,
        parameters: z.object({
          reportTypes: z
            .string()
            .describe(
              "...."
            ),
          question: z.string().describe("the user's question"),
        }),
        execute: async ({ question, reportTypes }) =>
          getReports(question, reportTypes),
      }),
    },
    experimental_toolCallStreaming: true,
  })

  return result.toAIStreamResponse()
}
```

**Comment by lgrammel** [comment]
I tried it locally and tool calling roundtrips with vertex seem to work. We changed `@google-cloud/vertexai` into a peer dependency in `v0.0.16` to allow you to upgrade it separately.

Which version of `@google-cloud/vertexai` are you using? (we are currently testing against `1.3.1`)

**Comment by Godrules500** [comment]
@lgrammel 
I'm getting this currently with @google-cloud/vertexai 1.4.1 and 1.6.0

```
⨯ ClientError: [VertexAI.ClientError]: got status: 400 Bad Request. {"error":{"code":400,"message":"Invalid value at 'contents[2].parts[0].function_response.response' (type.googleapis.com/google.protobuf.Struct), \"It is currently 25°F and Sunny in newnan!\"","status":"INVALID_ARGUMENT","details":[{"@type":"type.googleapis.com/google.rpc.BadRequest","fieldViolations":[{"field":"contents[2].parts[0].function_response.response","description":"Invalid value at 'contents[2].parts[0].function_response.response' (type.googleapis.com/google.protobuf.Struct), \"It is currently 25°F and Sunny in newnan!\""}]}]}}
    at throwErrorIfNotOK (post_fetch_processing.js:32:19)
    at processTicksAndRejections (task_queues.js:95:5)
    at async generateContentStream (generate_content.js:97:5)
    ...
    (stack trace continues)
```

**Comment by lgrammel** [comment]
@Godrules500 I've just tried out https://github.com/vercel/ai/blob/main/examples/ai-core/src/generate-text/google-vertex-tool-call.ts with both Vertex 1.4.1 and 1.6.0 - it works for me. Can you share the example that is failing for you?

**Comment by kn** [comment]
FWIW, I think this happens when passing `messages` into `streamText` where `messages contain assistant message with toolInvocations, so [your code snippet](https://github.com/vercel/ai/blob/main/examples/ai-core/src/generate-text/google-vertex-tool-call.ts) might not trigger it.

My code is:
```
  const result = await streamText({
    model: model,
    system,
    messages: convertToCoreMessages(messages),
    tools: {
        someTool: {
          description: 'Present list items for users to select.',
          parameters: z.object({
            options: z.array(z.string()),
          }),
          execute: async ({ options }) => {
            return options
          },
        }
    },
  })
```

This code works fine for the initial call where `messages` only contains a user message. Then the subsequent call fails when `messages` contains an assistant message with `toolInvocations`. So it's probably something to do with sending `toolInvocations` to Vertex.

Ex:
```
[
  {
    role: 'user',
    content: 'Testing',
  },
  { role: 'assistant', content: '', toolInvocations: [ [Object] ] },
  { role: 'user', content: 'Use option 1' }
]
```

**Comment by mauricekleine** [comment]
@lgrammel I think I found the issue - in those cases where the `execute` function returns an object, all is good. So your example works.

However, if you change the `return` in the `execute` function to `return 72 + Math.floor(Math.random() * 21) - 10` for example, it will fail. 

I'm encountering similar issues in my codebase, where I want to return a plain string from `execute` but then I receive the 400 bad request error. It only works if I return an object instead (e.g. `return { text: response }` instead of `return response`).

**Comment by metaloozee** [comment]
> @lgrammel I think I found the issue - in those cases where the `execute` function returns an object, all is good. So your example works.
> 
> However, if you change the `return` in the `execute` function to `return 72 + Math.floor(Math.random() * 21) - 10` for example, it will fail.
> 
> I'm encountering similar issues in my codebase, where I want to return a plain string from `execute` but then I receive the 400 bad request error. It only works if I return an object instead (e.g. `return { text: response }` instead of `return response`).

This is most likely the issue, even I tried out your approach and it worked for a simple JSON object like
```
{ "toolResult": "Hello World" }
```

But let's say I return an Array of Objects, the same error occurs.

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
This is a concrete provider serialization failure in existing tool-roundtrip behavior, subsequently addressed by a code fix.

### Evidence
- Vertex tool-call roundtrips return a 400 INVALID_ARGUMENT because tool results are serialized into an invalid Vertex `function_response` payload.
- Multiple commenters reproduced the failure when tools return strings, numbers, arrays, or arrays of objects; wrapping the result in an object avoids it.
- A maintainer confirmed that a fix landed on main for the upcoming 4.0 release.
